### DISCUSSION TITLE
How to handle XRootD error: [ERROR] Operation expired

### DISCUSSION THREAD

**Discussion by dcervenkov** [scikit-hep/uproot5#355] [Q&A]
When reading many files via XRootD, we see intermittent and seemingly non-deterministic "Operation expired" errors from XRootD. This might be an XRootD or an EOS issue, but I filed it here since we are experiencing it when using Uproot. I'd be interested to hear if you have ideas about how to troubleshoot this.

MWE to reproduce follows. Unfortunately, it requires access to LHCb files on EOS.

```py
import uproot

samples = [
    f"root://eoslhcb.cern.ch//eos/lhcb/grid/prod/lhcb/LHCb/Collision16/PIDCALIB.ROOT/00111823/0000/00111823_{i:08}_1.pidcalib.root"
    for i in range(1, 239)
]

for file in samples:
    print(f"Reading {file}")
    tree = uproot.open(file)["DSt_PiMTuple"]
```

```
Traceback (most recent call last):
  File "mwe_xrootd.py", line 12, in <module>
    tree = uproot.open(file)["DSt_PiMTuple"]
  File "/home/dc/.local/lib/python3.6/site-packages/uproot/reading.py", line 145, in open
    **options  # NOTE: a comma after **options breaks Python 2
  File "/home/dc/.local/lib/python3.6/site-packages/uproot/reading.py", line 553, in __init__
    file_path, **self._options  # NOTE: a comma after **options breaks Python 2
  File "/home/dc/.local/lib/python3.6/site-packages/uproot/source/xrootd.py", line 434, in __init__
    self._open()
  File "/home/dc/.local/lib/python3.6/site-packages/uproot/source/xrootd.py", line 440, in _open
    for x in uproot._util.range(self._num_workers)
  File "/home/dc/.local/lib/python3.6/site-packages/uproot/source/xrootd.py", line 440, in <listcomp>
    for x in uproot._util.range(self._num_workers)
  File "/home/dc/.local/lib/python3.6/site-packages/uproot/source/xrootd.py", line 88, in __init__
    self._xrd_error(status)
  File "/home/dc/.local/lib/python3.6/site-packages/uproot/source/xrootd.py", line 108, in _xrd_error
    status.message, self._file_path
OSError: XRootD error: [ERROR] Operation expired
in file root://eoslhcb.cern.ch//eos/lhcb/grid/prod/lhcb/LHCb/Collision16/PIDCALIB.ROOT/00111823/0000/00111823_00000057_1.pidcalib.root
```

It is not a problem with expired credentials, etc. - just rerunning the script will cause it to fail on a different file or not at all.

The issue occurs on multiple computers with very different setups.

uproot 4.0.7
xrootd 5.1.1

**Comment by chrisburr** [comment]
Interestingly I can't reproduce it on lxplus but I can when reading the data from outside of the CERN. Now to figure out why...

**Comment by chrisburr** [comment]
This isn't really an uproot problem (ROOT also had the same issue) but I'll summarise how I debugged it in case others run in to the same issue.

1. Repeating the test many times showed that it only ever crashed on a subset of the files mentioned in this issue
2. `eos fileinfo /eos/lhcb/...` showed that they all contained a replica that was on the same server
3. It was reported at https://ggus.eu/?mode=ticket_info&ticket_id=151899 and indeed there was a server with a problem that prevented it from being made visible outside the firewall.

**Comment by dcervenkov** [comment]
That's good news. Thanks for taking care of it @chrisburr!

**Comment by jpivarski** [comment]
Thanks, @chrisburr! To make sure your solution remains visible, I'll convert this into a discussion.

**Comment by ryuwd** [comment]
Advance apologies as this is not an 'answer', but adding a similar problem I have encountered in more recent versions of `uproot4` and `xrootd`, to the discussion.

I am also seeing `Operation expired` errors and `Invalid response` errors when using xrootd with `uproot4` 4.0.11 with `XRootDSource` as the `xrootd_handler`. I tried using `XRootDSource` again in my analysis code after seeing xrootd 5.2.0 was recently released, and some updates had been pushed to uproot4 as well (moving `XRootDSource` back to the default `xrootd_handler`.)

I'm reading chunks from trees across many files, streaming them from CERN to my institute, then processing using `dask` delayed (see the end of my post for the exact code.) I don't think `uproot4` is the problem, but when using XRootD vector reads (I use `xrootd=5.2.0` from conda-forge in my analysis environment) with `XRootDSource`, reads will occasionally fail with either `Operation expired`, `Invalid response`, and more rarely I am seeing `EPoll: bad file descriptor polling for events`

When I use `MultithreadedXRootDSource` the problems 'go away' and my jobs finish successfully. (Apologies for this vague analysis!)

LHCb-related: these ntuples come from the analysis production `XicpStStToXicpPiPi`
example URLs:
```
root://eoslhcb.cern.ch//eos/lhcb/grid/prod/lhcb/LHCb/Collision16/XICPSTSTTOXICPPIPI_DATA_2016_MAGUP.ROOT/00130129/0000/{id}.root

root://eoslhcb.cern.ch//eos/lhcb/grid/prod/lhcb/MC/2016/XICPSTSTTOXICPPIPI_MC.ROOT/00130499/0000/{id}.root
```

When I have some more time to debug this, I'll try to run this through again with `XRootDSource`, collect the tracebacks, and report them properly. Anyway, it seems for lots of XRootD root file streaming that `MultithreadedXRootDSource` is the potentially more stable option (although perhaps not ideal speed-wise?)

---

```python

def get_file_handler(filename):
    xrootd_src = filename.startswith("root://")
    if not xrootd_src:
        return {"file_handler": uproot.MultithreadedFileSource} # otherwise the memory maps overload available Vmem
    elif xrootd_src:
        # uncomment below for MultithreadedXRootDSource
        # return {"xrootd_handler": uproot.source.xrootd.MultithreadedXRootDSource}
    return {}

def load_root_file(tree, columns, total_file_parts=4):
    def get_tree_part(nentries, part, total_parts):
        part_len = nentries / total_parts
        entry_start = part * part_len
        entry_stop = (
            nentries if (total_parts - 1) == (part) else int(entry_start + part_len)
        )
        return int(entry_start), entry_stop

    def root_loader(filename, file_part):
        with uproot.open(filename, **get_file_handler(filename)) as rf:
            treeo = rf[tree]
            estart, estop = get_tree_part(
                treeo.num_entries, file_part, total_file_parts
            )
            df = rf[tree].arrays(
                columns, entry_start=estart, entry_stop=estop, library="pd"
            )
        assert (estop - estart) == df.shape[0]
        return df

    return root_loader

def load_root_as_dask_df(filenames, tree, columns, total_file_parts=1):
    meta = get_dtypes_root(filenames[0], tree, columns)
    loader = load_root_file(tree, columns, total_file_parts=total_file_parts)
    dfs = []

    for fn in filenames:
        for part in range(total_file_parts):
            dfs += [delayed(loader, pure=True, traverse=False)(fn, part)]

    df = dd.from_delayed(dfs, meta=meta)

    return df[columns]

```

**Comment by ellkay7** [comment]
I just wanted to report that we are currently having this issue using uproot  4.1.5. 
We are running code on lxplus and trying to access files both on the CERN eos and on a /pnfs file system for another institute. 
We get these OS errors, but I also see jobs getting killed when trying to read files using xrootd, as opposed to directly reading from the local eos mount. 

We are using uproot.iterate to loop over the files in chunks of defined sizes. I can try to whip together a MWE if needed. 
It seems especially strange in the case of the eos files, since we just try to read the exact same files that we have successfully analysed in the past in uproot but with the root://eosatlas... string prepended. 

I did try setting the file_handler to MultithreadedXRootDSource in iterate, but this didn't improve our situation (jobs were hanging and eventually killed after the first iteration)

@jpivarski is there a possibility that this needs to be upgraded from discussion to issue again?

**Reply by ryuwd** [reply]
Hi @special-kay-0,

If your files are on eos it might help to take the corresponding file where the `Operation expired` error is happening and try
```
eos info <eos path to file>
```
to narrow down the issue.

I have found that from time to time files become unavailable due to the servers at CERN providing the file going offline.  You can then report that to someone who manages the EOS / grid storage / data management facilities for your collaboration

**Reply by ellkay7** [reply]
Hi @ryuwd 

Yes, we do deal with intermittent eos access issues. However, there is a clear difference in uproot between accessing them directly via the eos path w.r.t. accessing  with xrootd. 
We have come to expect the occasional problem accessing a given file, but with this xrootd access we are systematically seeing the OSError or hanging file access and excess memory use leading to jobs being killed.

**Reply by ryuwd** [reply]
Just learned recently that one can run into XRootD `Operation expired` errors if authentication is not valid. For example,

 - invalid (or no) kerberos ticket
 - invalid (or no) eos token provided
 - invalid x509 auth
 - some other server issue
 
 The request would time out before the remote has a chance to check authentication and return the more relevant error, hence the cryptic `Operation expired`. You of course wouldn't see such an error with the `/eos` mount, because the authentication is usually transparent to the user in that case.

**Reply by dcervenkov** [reply]
Interesting - when there's no Kerberos ticket, I get `FileNotFound` with "Server responded with error: [3010]" in its arguments.

**Comment by msnuaa** [comment]
Hi, all, I am suffering the same issue of `Operation expired` at one of the remote site, and only for some of the files on the same site.
I am using root 6.16 and uproot 4.3.5. Is there a proper way to get rid of this issue?
Thanks in advance!

**Reply by gabhijith** [reply]
I am facing a similar issue on cms-lpc.

**Comment by amishgog** [comment]
Hey All. 

I am facing this issue on lxplus. I have verified that I have valid Kerberos ticket and valid x509 proxy. My analysis had been working fine a while ago but I have been facing this issue consistently for the past few days. 

I am using `apply_to_fileset` method to run the analysis on some root files organised in a fileset. If this is an issue with a certain root file, the error provides no information on that. 

Any help would be appreciated. 

Cheers

**Reply by ikrommyd** [reply]
So...the data delivery infrastructure is fragile overall. This error is not coming from `uproot` directly but from xrootd. What I typically do is use `allow_read_errors_with_report` in `apply_to_fileset` to catch these errros so that the whole workflow doesn't crash. See here: https://github.com/scikit-hep/coffea/blob/685e091d818d47b8c8960bf8e31823c090127402/src/coffea/dataset_tools/apply_processor.py#L112-L113. You can also pass in a tuple of errors to catch like `(OSError, ValueError)` instead of just `True` which will just catch `OSError`. Then reading errors will be caught and put in a report and then you can see what the error was for each chunk of data and also get a smaller fileset of all the smaller steps using https://github.com/scikit-hep/coffea/blob/685e091d818d47b8c8960bf8e31823c090127402/src/coffea/dataset_tools/manipulations.py#L247-L270.

**Reply by ikrommyd** [reply]
What I typically do is that for simulation I don't care. If I miss a few chunks of data, it's fine.
For data where I want all my data I either rerun with the failed chunks of the previous run until all of the chunks succeed and merge the results, or copy all the files locally so that I don't get xrootd errors due to xrootd reading, or keep track of which luminosity blocks I was able to successfully run over in order to calculate the proper luminosity to normalize my simulation to in case I miss a few chunks of data events.

**Reply by amishgog** [reply]
Hey @ikrommyd

Thanks for the response. In my uproot options I have taken `UPROOT_OPTIONS = {"allow_read_errors_with_report": True, "skipbadfiles": True, "timeout": 120}`. 

The weird thing is that this analysis fully worked before last week. But the whole last week it consistently returned this error. I doubt it is related to the timeline but still weird. 

Also while I was creating the fileset. I explicitly skipped the files that return an OSError.

**Reply by ikrommyd** [reply]
Yeah the file access errors are random very often. It doesn't mean that something that didn't error when you were preprocessing, will never error in the future. So you do use `allow_read_errors_with_report` and still hit the actual error and your workflow crashes?

**Reply by amishgog** [reply]
Yeah. Also one thing I should have pointed out earlier is that I get an OSError and not an XRootD error. 

I am attaching a snippet of the error below:

```
return XRootDFile(
           ^^^^^^^^^^^
  File "/cvmfs/sft.cern.ch/lcg/views/LCG_107/x86_64-el9-gcc13-opt/lib/python3.11/site-packages/fsspec_xrootd/xrootd.py", line 768, in __init__
    raise OSError(f"File did not open properly: {status.message}")
OSError: File did not open properly: [ERROR] Operation expired
```

**Reply by ikrommyd** [reply]
That an `OSError` coming from xrootd. It's being raised in `fsspec_xrootd/xrootd.py`. That's weird....the report should have caught that.

**Reply by amishgog** [reply]
Do you think it is originating from a particular root file or is there an issue with the analysis code?
